### PR TITLE
Bag of words updates

### PR DESCRIPTION
**Pull request by NikhilTilak**
- bow now removes proper nouns (Character names, City names etc.)

- lemmatization before vectorization implemented.

- ability to remove too frequent or too infrequent words added to mimic the CountVectorizer
parameters max_df and min_df.

- Model can now be retrained.

- Added a flag called  use_passed_bow which decides if the passed bag of words will be use.

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -135,38 +139,74 @@ class BOWFeatures(BaseEstimator, TransformerMixin):
...
-        if self.bow =={}:
-            self.build_bow(corpus = text_frame['text'].values)
+        
+        if not self.use_passed_bow:
+            self.build_bow(corpus = text_frame['text'])
+            
+        else:
+            if self.bow =={}:
+                print("Looks like you forgot to pass a Bag of words.")
+                print("If you want to build a bag instead, set use_passed_bag=False .")
```
Comment (emapple) [inline src/model.py]: This should probably raise a `ValueError` instead of printing out the warning

File: src/model.py [hunk 2]
```diff
@@ -175,30 +215,130 @@ def build_bow(self, corpus):
...
+        
+        
+        #--------------------------- identify and remove all proper nouns ------------------------
+        
+        if self.include_proper_nouns:
+            proper_nouns = []
+        
+        else:
+            proper_nouns= self.proper_nouns_in_corpus(corpus) 
+            proper_nouns = [word.lower() for word in proper_nouns]
```
Comment (emapple) [inline src/model.py:266]: This line can accidentally remove words that aren't proper nouns. For example (and this just happened to be my test case!):

```
from src.model import BOWFeatures
bow = BOWFeatures()
bow.fit(['This is a test. Carson City is a city.'])

# this yields {'test': 0}, because city was spuriously removed
print(bow.bow)
```

I assume you added `lower` to prevent missing any. So you'll have to decide whether you'd rather accidentally keep some proper nouns, or accidentally remove non-proper nouns.

File: src/model.py [hunk 3]
```diff
@@ -175,30 +215,130 @@ def build_bow(self, corpus):
...
+    def proper_nouns_in_corpus(self, corpus: pd.Series):
+        
+        """ collects all proper nouns in the corpus and returns them as a list"""
+
+        def find_proper_nouns(text):
+            """ A local helper function """
+            bt = BookText(rawtext=text)
+            bt.clean(lemmatize=True, deromanize=True,inplace=True)
+            sentences = bt.tokenize(on = 'sent') #all sentences in one row
+            tagged_sentences = [pos_tag(sentence.split()) for sentence in sentences]  # tagged sentences
```
Comment (emapple) [inline src/model.py:291]: I'm not sure it matters, but `nltk.pos_tag_sents` is allegedly a lot faster than `pos_tag`

**Review by emapple — CHANGES_REQUESTED** [review]

**Comment by emapple** [comment]
Overall, there are some solid improvements here. Refit works like a charm, and the other parameters seem to work well.

File: src/model.py (hunk 1 above)
Comment (NikhilTilak) [reply src/model.py]: OK. That's reasonable.

File: src/model.py (hunk 2 above)
Comment (NikhilTilak) [reply src/model.py:266]: This is one of those corner cases which is difficult to deal with in a generalized way. The list of proper nouns is in title case while the whole bag is in lower case, so in order to compare them they have to be converted to lower case at some point.

If instead we decide to make the bag of words preserve case, then words which appear at the beginning of sentences will be counted as separate words than the same word which appears in the middle. That would open a whole other can of worms.

In conclusion, I'm okay with accidentally removing some common nouns.

Comment (emapple) [reply src/model.py:266]: Alright, that seems reasonable

File: src/model.py (hunk 3 above)
Comment (NikhilTilak) [reply src/model.py:291]: I looked into it and tried incorporating it into the code. 
Unfortunately because of the way bt.tokenize returns the sentences it cannot be easily implemented without adding an additional data massaging step. This step completely defeats the tiny gains in speed from switching to pos_tag_sents.

I  am choosing to keep it as is.

File: src/model.py [hunk 4]
```diff
@@ -196,8 +196,7 @@ def fit(self, X, y=None):
             
         else:
             if self.bow =={}:
-                print("Looks like you forgot to pass a Bag of words.")
-                print("If you want to build a bag instead, set use_passed_bag=False .")
+                raise ValueError
```
Comment (emapple) [inline src/model.py]:
```suggestion
                raise ValueError('Must pass a bag of words if use_passed_bow is True')
```

**Review by emapple — CHANGES_REQUESTED** [review]

**Review by emapple — APPROVED** [review]

**Comment by diptanil-roy** [comment]
Works like a charm. Tested.

**Review by diptanil-roy — APPROVED** [review]
